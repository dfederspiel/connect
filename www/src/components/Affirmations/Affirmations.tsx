import { useMutation, useQuery } from '@apollo/client';
import { GetUsersDocument, SendAffirmationDocument } from '../../gql/graphql';

const Affirmations = () => {
  const { loading, error, data } = useQuery(GetUsersDocument);
  const [sendAffirmation] = useMutation(SendAffirmationDocument);

  return (
    (loading && <div>Loading...</div>) ||
    (error && <div>Error! {error.message}</div>) ||
    (data?.users && (
      <div className="users-list">
        {data.users.map(
          (user) =>
            user && (
              <div data-testid="user" key={user.id}>
                <button
                  type="button"
                  onClick={() => {
                    sendAffirmation({
                      variables: {
                        userId: user.id,
                      },
                    });
                  }}
                >
                  Send
                </button>
                <span>{user.id}</span>
              </div>
            ),
        )}
      </div>
    )) || <>Unplanned Code Path</>
  );
};

export default Affirmations;
