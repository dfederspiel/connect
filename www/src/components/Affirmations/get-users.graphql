query GetUsers {
  users {
    id
    email
  }
}
