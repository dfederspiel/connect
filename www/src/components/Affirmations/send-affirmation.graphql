mutation sendAffirmation($userId: ID!) {
  sendAffirmation(userId: $userId) {
    from
    to
  }
}
