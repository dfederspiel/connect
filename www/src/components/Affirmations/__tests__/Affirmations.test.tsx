import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { User } from '@prisma/client';
import { render, cleanup, act, fireEvent } from '@testing-library/react';
import Affirmations, { GET_USERS, SEND_AFFIRMATION } from '../Affirmations';

let defaultMock = {
  request: {
    query: GET_USERS,
  },
  result: {
    data: {
      users: [
        {
          id: 1,
          domain: 'domain',
          email: '',
        },
      ] as User[],
    },
  },
} as MockedResponse;

const affirmationMock = {
  request: {
    query: SEND_AFFIRMATION,
    variables: {
      userId: 1,
    },
  },
  result: {
    data: {},
  },
} as MockedResponse;

afterEach(cleanup);

describe('the affirmations component', () => {
  const wrapper = ({ children }: any) => (
    <MockedProvider mocks={[defaultMock, affirmationMock]}>{children}</MockedProvider>
  );
  it('initializes with a loading state', () => {
    const { asFragment } = render(<Affirmations />, { wrapper });
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays a list of users', async () => {
    const { asFragment, findAllByTestId } = render(<Affirmations />, { wrapper });
    const users = await findAllByTestId('user');
    expect(users.length).toBeGreaterThan(0);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('and when the button is clicked', () => {
    it('it triggers a send affirmation call', async () => {
      const { findByRole } = render(<Affirmations />, { wrapper });
      const affirmButton = await findByRole('button');
      act(() => {
        fireEvent.click(affirmButton);
      });
    });
  });

  it('displays an error when the query fails', async () => {
    defaultMock = {
      request: {
        query: GET_USERS,
      },
      error: new Error('An error occurred'),
    };

    const wrapper = ({ children }: any) => (
      <MockedProvider mocks={[defaultMock, affirmationMock]}>{children}</MockedProvider>
    );
    const { findByText } = render(<Affirmations />, { wrapper });
    await findByText(`Error! An error occurred`);
  });
});
