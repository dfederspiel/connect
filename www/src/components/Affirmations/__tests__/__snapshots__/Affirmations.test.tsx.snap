// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the affirmations component displays a list of users 1`] = `
<DocumentFragment>
  <div
    data-testid="user"
  >
    <button
      type="button"
    >
      Send
    </button>
    <div>
      1
    </div>
  </div>
</DocumentFragment>
`;

exports[`the affirmations component displays an error when the query fails 1`] = `
<DocumentFragment>
  <div>
    Error! An error occurred
  </div>
</DocumentFragment>
`;

exports[`the affirmations component initializes with a loading state 1`] = `
<DocumentFragment>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;
